torch/lib/THD/base/data_channels/GlooCache.hpp

```diff
@@ -0,0 +1,238 @@
+#pragma once
+
+#include "../ChannelUtils.hpp"
+#include "../DataChannel.hpp"
+
+#include "gloo/algorithm.h"
+#include "gloo/allgather_ring.h"
+#include "gloo/allreduce_ring.h"
+#include "gloo/barrier_all_to_all.h"
+#include "gloo/broadcast_one_to_all.h"
+#include "gloo/rendezvous/context.h"
+#include "gloo/rendezvous/store.h"
+#include "gloo/rendezvous/prefix_store.h"
+
+#include <cstdint>
+#include <memory>
+#include <tuple>
+#include <vector>
+
+
+MAKE_HASHABLE(
+  SINGLE_ARG(std::tuple<thd::CollectiveType, THDGroup, std::size_t, std::size_t,
```

Comment:
Maybe it would be possible to move it below and use `GlooCache::key_type`. Otherwise `using`